src/api/app/models/project.rb

```diff
@@ -1043,19 +1043,11 @@ def expand_maintained_projects
   end
 
   def packages_from_linked_projects
-    linkedprojects.each do |project|
-      next if project.linked_db_project.nil?
-      project.linked_db_project.packages_from_linked_projects.each do |name, prj_id|
-        unless p_map[name]
-          packages << [name, prj_id]
-          p_map[name] = 1
-        end
-      end
-    end
-  end
-
-  def linked_packages
-    expand_all_projects({}, false).reject{|project| project == self}.map {|project| project.packages.to_a }.flatten
+    all_packages = Package.
+        joins('LEFT OUTER JOIN linked_projects ON packages.project_id = linked_projects.linked_db_project_id').
+        where('linked_projects.db_project_id = ? AND name NOT IN (?)', id, packages.pluck(:name)).
+        order('linked_projects.position ASC, packages.name ASC')
+    all_packages.to_a.uniq { |p| p.name }
```

Comment:
Yeah, already tried it but unfortunately the database uniq doesn't preserve the ordering. We can talk later if you have a better idea for the query!